src/jit/morph.cpp

```diff
@@ -926,130 +926,68 @@ fgArgInfo::fgArgInfo(GenTreeCall* newCall, GenTreeCall* oldCall)
-        newArgs   = newArgs->GetNext();
-        oldArgs   = oldArgs->GetNext();
-
-        fgArgTabEntry* oldArgTabEntry = nullptr;
-        fgArgTabEntry* newArgTabEntry = nullptr;
+        argTable = new (compiler, CMK_fgArgInfoPtrArr) fgArgTabEntry*[argTableSize];
 
-        for (unsigned inx = 0; inx < argTableSize; inx++)
+        // Copy the old arg entries
+        for (unsigned i = 0; i < argTableSize; i++)
         {
-            oldArgTabEntry = oldArgTable[inx];
+            argTable[i] = new (compiler, CMK_fgArgInfo) fgArgTabEntry(*oldArgInfo->argTable[i]);
+        }
 
-            if (oldArgTabEntry->use == oldParent)
+        // The copied arg entries contain pointers to old uses, they need
+        // to be updated to point to new uses.
+        if (newCall->gtCallThisArg != nullptr)
+        {
+            for (unsigned i = 0; i < argTableSize; i++)
             {
-                assert((oldParent == nullptr) == (newParent == nullptr));
-
-                // We have found the matching "parent" field in oldArgTabEntry
-
-                newArgTabEntry = new (compiler, CMK_fgArgInfo) fgArgTabEntry;
-
-                // First block copy all fields
-                //
-                *newArgTabEntry = *oldArgTabEntry;
-
-                // Then update all GenTree* fields in the newArgTabEntry
-                //
-                newArgTabEntry->use = newParent;
-
-                if (oldArgTabEntry->lateUse != nullptr)
+                if (argTable[i]->use == oldCall->gtCallThisArg)
                 {
-                    newArgTabEntry->lateUse = nullptr;
-                    scanRegArgs             = true;
+                    argTable[i]->use = newCall->gtCallThisArg;
+                    break;
                 }
-
-                // Now initialize the proper element in the argTable array
-                //
-                argTable[inx] = newArgTabEntry;
-                break;
             }
         }
-        // We should have found the matching oldArgTabEntry and created the newArgTabEntry
-        //
-        assert(newArgTabEntry != nullptr);
-    }
 
-    if (scanRegArgs)
-    {
-        newArgs = newCall->gtCallLateArgs;
-        oldArgs = oldCall->gtCallLateArgs;
+        GenTreeCall::UseIterator newUse    = newCall->Args().begin();
+        GenTreeCall::UseIterator newUseEnd = newCall->Args().end();
+        GenTreeCall::UseIterator oldUse    = oldCall->Args().begin();
+        GenTreeCall::UseIterator oldUseEnd = newCall->Args().end();
 
-        while (newArgs != nullptr)
+        for (; newUse != newUseEnd; ++newUse, ++oldUse)
         {
-            /* Get hold of the next argument values for the oldCall and newCall */
-
-            newParent = newArgs;
-            newArgs   = newArgs->GetNext();
-
-            oldParent = oldArgs;
-            oldArgs   = oldArgs->GetNext();
+            for (unsigned i = 0; i < argTableSize; i++)
+            {
+                if (argTable[i]->use == &(*oldUse))
```

Comment:
How much time do we spend in this quadratic loop? 

Nit: I do not like ` &(*oldUse)` with hiding castings.

Reply:
Luckily cloning calls isn't exactly common. The only case I know of is loop hoisiting and that only works with a few "pure" helpers that have 2 arguments.

I don't like `&(*oldUse)` but I don't understand what do you mean by hiding casts. The only way to avoid this seems to be to avoid using iterators but I'd like to avoid relying on the fact that the arg list is a linked list. It should be an array and using iterators makes changing it easier. Assuming we ever figure out how to deal with the few cases that require as to insert new arguments into this arg list...

Reply:
`oldUse`'s type is `UseIterator `,
`*oldUse`'s type is 'Use&' (because of operator overload `Use& operator*() const`),
`'&(*oldUse)`'s type is `Use*`.

with inlining, it looks like `&*(oldUse->m_use)`, maybe we can add `Use* UseIterator::GetUse() const` and call it here and below?